### PR TITLE
mgr/snap_schedule: fix db connection concurrent usage

### PR DESCRIPTION
**Pull request by mchangir**
Serialize access to DB connection to avoid transaction aborts due to
concurrent use.

Fixes: https://tracker.ceph.com/issues/52642
Signed-off-by: Milind Changire <mchangir@redhat.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Review by batrick — CHANGES_REQUESTED** [review]
Have you reproduced the issue, then? Can you create a QA test that does it? I imagine it would possible with several 1m snapshot schedules?

**Comment by mchangir** [comment]
> Have you reproduced the issue, then? Can you create a QA test that does it? I imagine it would possible with several 1m snapshot schedules?

Yes, I could successfully reproduce the issue on my vstart.sh cluster with 1M snapshot schedules.
The code is now working without deadlocks and with correct statistics in the `status` dump.
I'll add a QA Test Case commit to the PR.

fyi:
m => month(s)
M => minute(s) - needs `$ ceph config set mgr.x mgr/snap_schedule/allow_m_granularity true`
I think the `m` in the `allow_m_granularity` should've been capitalized as well ... just for emphasis

**Comment by mchangir** [comment]
> @mchangir ping?

@vshankar 
PR has been updated.
Please check.

**Review by vshankar — CHANGES_REQUESTED** [review]
```
2021-10-14T12:39:48.248+0000 7f5dd022ee80 10 mgr[py] Computed sys.path '/usr/share/ceph/mgr:/usr/local/lib64/python3.6/site-packages:/usr/local/lib/python3.6/site-packages:/usr/lib64/python3.6/site-packages:/usr/lib/python3.6/site-packages:/usr/lib64/python36.zip:/usr/lib64/python3.6:/usr/lib64/python3.6:/usr/lib64/python3.6/lib-dynload'
2021-10-14T12:39:48.319+0000 7f5dd022ee80 -1 mgr[py] Module not found: 'snap_schedule'
2021-10-14T12:39:48.319+0000 7f5dd022ee80 -1 mgr[py] Traceback (most recent call last):
  File "/usr/share/ceph/mgr/snap_schedule/__init__.py", line 11, in <module>
    from .module import Module
  File "/usr/share/ceph/mgr/snap_schedule/module.py", line 10, in <module>
    from .fs.schedule_client import SnapSchedClient
  File "/usr/share/ceph/mgr/snap_schedule/fs/schedule_client.py", line 6
    from __future__ import annotations
                                     ^
SyntaxError: future feature annotations is not defined

2021-10-14T12:39:48.320+0000 7f5dd022ee80 -1 mgr[py] Class not found in module 'snap_schedule'
2021-10-14T12:39:48.320+0000 7f5dd022ee80 -1 mgr[py] Error loading module 'snap_schedule': (2) No such file or directory
```

**Comment by mchangir** [comment]
> > [annotations](https://docs.python.org/3/library/__future__.html#module-__future__) | optional in: 3.7.0b1 | mandatory in: 3.11
> 
> teuthology uses 3.6 (atleast in rhel/centos). any possibility of not using this feature in the module?

done
I've used an alternative way for type hinting

**Review by vshankar — DISMISSED** [review]
@mchangir 

```
2021-10-19T09:26:16.105 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2021-10-19T09:26:16.105 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_3173226cc2328162242c7b05c648747279967032/qa/tasks/cephfs/test_snap_schedules.py", line 292, in test_multi_snap_schedule
2021-10-19T09:26:16.106 INFO:tasks.cephfs_test_runner:    self.assert_if_not_verified()
2021-10-19T09:26:16.106 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_3173226cc2328162242c7b05c648747279967032/qa/tasks/cephfs/test_snap_schedules.py", line 118, in assert_if_not_verified
2021-10-19T09:26:16.107 INFO:tasks.cephfs_test_runner:    self.assertListEqual(self.create_cbks, [])
2021-10-19T09:26:16.107 INFO:tasks.cephfs_test_runner:AssertionError: Lists differ: [<function TestSnapSchedules.test_multi_sn[50 chars]c80>] != []
2021-10-19T09:26:16.107 INFO:tasks.cephfs_test_runner:
2021-10-19T09:26:16.108 INFO:tasks.cephfs_test_runner:First list contains 1 additional elements.
2021-10-19T09:26:16.108 INFO:tasks.cephfs_test_runner:First extra element 0:
2021-10-19T09:26:16.108 INFO:tasks.cephfs_test_runner:<function TestSnapSchedules.test_multi_snap_schedule.<locals>.verify_added_1 at 0x7fd440787c80>
2021-10-19T09:26:16.108 INFO:tasks.cephfs_test_runner:
2021-10-19T09:26:16.109 INFO:tasks.cephfs_test_runner:- [<function TestSnapSchedules.test_multi_snap_schedule.<locals>.verify_added_1 at 0x7fd440787c80>]
2021-10-19T09:26:16.109 INFO:tasks.cephfs_test_runner:+ []
```

I can reproduce this locally too. Looks like the scheduled snapshots (at times) are not getting taken.

See: https://pulpito.ceph.com/vshankar-2021-10-19_08:51:42-fs-wip-vshankar-testing-20211019-105809-distro-default-smithi/

**Comment by mchangir** [comment]
@vshankar rebased sources
[Teuthology run: 1 Failed; 1 Passed](https://pulpito.ceph.com/mchangir-2021-11-11_07:17:38-fs-wip-mgr_snap_schedule-fix-db-connection-concurrent-usage-distro-default-smithi/)

failed with mon warning:

    mon.a (mon.0) 376 : cluster [WRN] Health check failed: Reduced data availability: 1 pg peering (PG_AVAILABILITY)

**Comment by vshankar** [comment]
> ```
> mon.a (mon.0) 376 : cluster [WRN] Health check failed: Reduced data availability: 1 pg peering (PG_AVAILABILITY)
> ```

Maybe this just needs to be added to ignore-list?

**Comment by mchangir** [comment]
> > ```
> > mon.a (mon.0) 376 : cluster [WRN] Health check failed: Reduced data availability: 1 pg peering (PG_AVAILABILITY)
> > ```
> 
> Maybe this just needs to be added to ignore-list?

How do we tell teuthology to ignore this log if it is found after test executions complete successfully ?

**Comment by vshankar** [comment]
> How do we tell teuthology to ignore this log if it is found after test executions complete successfully ?

search for "log-ignorelist" under qa directory.

**Comment by mchangir** [comment]
> > How do we tell teuthology to ignore this log if it is found after test executions complete successfully ?
> 
> check "log-ignorelist" under qa directory.

I know the **log-ignorelist** part; I just don't know if that happens conditionally after the tests or in general throughout the run
Also, do I add the log ignore part for just these runs or permanently in one of the yamls under the ceph source tree ?

**Comment by mchangir** [comment]
@vshankar 
[Teuthology runs passed by temporarily ignoring pg peering warning](https://pulpito.ceph.com/mchangir-2021-11-11_13:48:33-fs-wip-mgr_snap_schedule-fix-db-connection-concurrent-usage-distro-default-smithi/)

**Comment by batrick** [comment]
Next time review:

https://tracker.ceph.com/projects/cephfs/wiki/Master

You would find:

https://tracker.ceph.com/issues/52624

**Comment by mchangir** [comment]
> Next time review:
> 
> https://tracker.ceph.com/projects/cephfs/wiki/Master
> 
> You would find:
> 
> https://tracker.ceph.com/issues/52624

ack

File: qa/suites/fs/functional/tasks/snap-schedule.yaml [hunk 1]
```diff
@@ -15,6 +15,7 @@ overrides:
       - is full \(reached quota
       - POOL_FULL
       - POOL_BACKFILLFULL
+      - pg peering \(PG_AVAILABILITY\)
```
Comment (batrick) [inline qa/suites/fs/functional/tasks/snap-schedule.yaml:18]: So the failure is annoying but it represents a real bug we should not squelch. Just document that it's a known problem when it shows up in qa results.

Let's remove this ignorelist addition.

**Review by batrick — CHANGES_REQUESTED** [review]

File: qa/suites/fs/functional/tasks/snap-schedule.yaml (hunk 1 above)
Comment (mchangir) [reply qa/suites/fs/functional/tasks/snap-schedule.yaml:18]: okay, I'll remove it from the ignorelist

**Comment by vshankar** [comment]
> Next time review:
> 
> https://tracker.ceph.com/projects/cephfs/wiki/Master

Somehow I missed seeing its already seen... sorry!

**Comment by vshankar** [comment]
https://pulpito.ceph.com/vshankar-2021-11-19_09:32:57-fs-wip-vshankar-testing-20211119-101529-testing-default-smithi/

**Comment by vshankar** [comment]
LGTM.

File: src/pybind/mgr/snap_schedule/fs/schedule.py [hunk 2]
```diff
@@ -239,9 +239,9 @@ def list_schedules(cls,
             else:
                 c = db.execute(cls.PROTO_GET_SCHEDULES + ' path = ?',
                                (f'{path}',))
-        return [cls._from_db_row(row, fs) for row in c.fetchall()]
+            return [cls._from_db_row(row, fs) for row in c.fetchall()]
 
-    INSERT_SCHEDULE = '''INSERT INTO
+    INSERT_SCHEDULE = '''INSERT OR IGNORE INTO
```
Comment (batrick) [inline src/pybind/mgr/snap_schedule/fs/schedule.py:244]: Why do we need this change?

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/snap_schedule/fs/schedule.py (hunk 2 above)
Comment (mchangir) [reply src/pybind/mgr/snap_schedule/fs/schedule.py:244]: As far as I could find docs on, without that additional clause, the python code dumps a `Traceback` while attempting to add a non unique row. The Python exception is expected as per design.
However, the `Traceback` is dumped even when there's an `except` clause to catch this every exception. This was the weird part.

I'm not sure if the traceback an artifact of Python 3.9 on my laptop when I started testing this piece of code. Teuthology systems run Python 3.6.
(I'm now running Python 3.10 after upgrade to Fedora 35)

I'm also not sure how a SQL DML statement can disregard the actual code in place and force a Traceback to be emitted by the Python run-time.

Comment (batrick) [reply src/pybind/mgr/snap_schedule/fs/schedule.py:244]: AFAICT, the code already considers the possibility of an exception thrown for an duplicate row. See the `except sqlite3.IntegrityError` below.

Comment (mchangir) [reply src/pybind/mgr/snap_schedule/fs/schedule.py:244]: The changes seemed to pass teuthology tests. So I presumed all's well.
I'll run a few tests independent of cephfs and see how the `OR IGNORE` clause interacts with `except sqlite3.IntegrityError` and get back here.

Comment (mchangir) [reply src/pybind/mgr/snap_schedule/fs/schedule.py]: [The original code without `IGNORE` clause works on teuthology](https://pulpito.ceph.com/mchangir-2021-11-24_04:15:48-fs-wip-mchangir-mgr-snap_schedule-fix-db-connection-concurrent-usage-distro-default-smithi/)

Oddly, the code works as expected by throwing and catching the `sqlite3.IntegrityError` exception in a standalone test app as well. But not in a `vstart_runner.py` environment on my laptop. This is really weird.

If somebody is up to it, please try on a Python 3.10 and `vstart_runner.py` environment.
I've attempted without success on a CentOS 8.3 virtual machine as well. Here, `vstart_runner.py` bails out when attempting to mount `/sys/fs/fuse/connections`.

I've now revered back to the original.

**Review by batrick — APPROVED** [review]

**Comment by vshankar** [comment]
https://tracker.ceph.com/projects/cephfs/wiki/Master#2022-Jan-03

**Pull request merged**